### ISSUE TITLE
Toast message in case of wrong credentials 

### ISSUE THREAD

**Issue by PratyushSingh07** [openMF/self-service-app#1986]
**Summary:**

If a user enters a wrong username or password then currently the toast being displayed is `Inavalid authentication details passed in api request` . I believe this should be changed to something more meaningful to the user such as `Incorrect username or password ` or  `Unable to authenticate user`

**Steps to reproduce:**

- enter either a wrong username or a wrong password at the time of login

**Device and Android version:**
Device : Samsung M11
Android : 11

**Screenshots:**
<table>
 <tr>
 <td><img src="https://user-images.githubusercontent.com/90026952/225122061-31944e1a-11cc-44ec-a1fd-2a20651c4bee.jpeg" width=250 height=480></td>
</table>

**Comment by PratyushSingh07** [comment]
@jawidMuhammadi do you think we need to change this toast? If yes then I can change the toast to something more comprehensible to a user

**Comment by oyehanif** [comment]
@jawidMuhammadi  @PratyushSingh07 do you think we need to  change with **set Error with required focus**. this look good in case of error.

**Comment by jawidMuhammadi** [comment]
@PratyushSingh07, yes, let's change the message to `Incorrect User Name or Password` and we would require translation for all the supported languages in the app. And yes @oyehanif, that is a good idea and I think using set error looks good if we  use it when checking for some validation before calling the login API like if the password is at least 6 characters and if the user name is not empty (Do we have those validations implemented in MifosMobile, could you please confirm?) and also we could the following if is not implemented yet:
- When the login button is clicked, then hide the keyboard
- Show the progress bar when authentication is in progress
- Wrong credentials? Show the message on a SnackBar instead of a Toast.

**Comment by PratyushSingh07** [comment]
> @PratyushSingh07, yes, let's change the message to `Incorrect User Name or Password` and we would require translation for all the supported languages in the app. And yes @oyehanif, that is a good idea and I think using set error looks good if we use it when checking for some validation before calling the login API like if the password is at least 6 characters and if the user name is not empty (Do we have those validations implemented in MifosMobile, could you please confirm?) and also we could the following if is not implemented yet:
> 
> * When the login button is clicked, then hide the keyboard
> * Show the progress bar when authentication is in progress
> * Wrong credentials? Show the message on a SnackBar instead of a Toast.

i can do these changes & I've already done some of them locally , so can i start working on rest of it and then open a PR?

**Comment by jawidMuhammadi** [comment]
Hey @PratyushSingh07, I would say let's label it as a `good first issue`, it is going to be a good start for new contributors who have not created any pull requests yet :)

**Comment by PratyushSingh07** [comment]
> Hey @PratyushSingh07, I would say let's label it as a `good first issue`, it is going to be a good start for new contributors who have not created any pull requests yet :)

ok @jawidMuhammadi

**Comment by RobustMj** [comment]
Although i am new can i try to solve this issue?

**Comment by jawidMuhammadi** [comment]
@RobustMj Yes, please go ahead

**Comment by RobustMj** [comment]
> @RobustMj Yes, please go ahead
I need some help am unable to find the toast or .json file which contains "Invalid authentication details were passed in api request" .

**Comment by gururani-abhishek** [comment]
Hey @RobustMj, I ran the application, and in the Logcat(attaching the screenshot below) found that this Toast text("Invalid authentication details were passed in api request") is being returned as the response body in OkHttp response when either of "username" and "password" are wrong.

<img width="1407" alt="Screenshot 2023-03-18 at 15 51 25" src="https://user-images.githubusercontent.com/59947871/226100460-85a487ca-c6c6-47d8-b03a-ef7916344f13.png">

If you want to change the toast message to a custom error text(what Jawid suggested), you need to put a "if"condition when the OkHttp exception code is = 401 and then have to provide your own custom error text. 

These points might help you to solve this issue : 
1. In LoginActivity.kt inside onLoginClicked(), "username" and "password" are passed to login() of LoginPresenter.kt class.
2. Inside login() in try-catch block, you can put an "if" condition for the response code = 401 which corresponds to invalid username and password exception.
eg : if(e.code == 401) mvpView?.showMessage("Invalid username or password"), I'm just hardcoding it right now but you should put an entry of this error in string.xml and also translate it to other languages. 

I'm looping @jawidMuhammadi, so that whenever he finds time he can confirm this : )

@RobustMj Tag me if you face any issue/s while dealing with this, I'll be more than happy to help.

**Comment by gururani-abhishek** [comment]
> @PratyushSingh07, yes, let's change the message to `Incorrect User Name or Password` and we would require translation for all the supported languages in the app. And yes @oyehanif, that is a good idea and I think using set error looks good if we use it when checking for some validation before calling the login API like if the password is at least 6 characters and if the user name is not empty (Do we have those validations implemented in MifosMobile, could you please confirm?) and also we could the following if is not implemented yet:
> 
> * When the login button is clicked, then hide the keyboard
> * Show the progress bar when authentication is in progress
> * Wrong credentials? Show the message on a SnackBar instead of a Toast.

Hey @jawidMuhammadi, I was reading our codebase and I found that these validations are implemented in our codebase. In the class LoginPresenter.kt  there's a function called ifCredentialsValid() which deals with all these validations.

**Comment by PratyushSingh07** [comment]
Hi @gururani-abhishek that wont be needed all @RobustMj needs to do is change the line 110 of LoginActivity.kt with an appropriate message in all the languages. (Use a Toaster and do not hardcode the string )
However even if you make these changes and allow support for multiple languages you won't be able to check the Toaster message when you switch to any other language as explained in the issue #1997 .

**Comment by RobustMj** [comment]
@gururani-abhishek thanks for the insight actually before posting i did debug the app to find exactly were the issue was.

**Comment by RobustMj** [comment]
> Hi @gururani-abhishek that wont be needed all @RobustMj needs to do is change the line 110 of LoginActivity.kt with an appropriate message in all the languages. (Use a Toaster and do not hardcode the string ) However even if you make these changes and allow support for multiple languages you won't be able to check the Toaster message when you switch to any other language as explained in the issue #1997 .

i added the toast for each of those error "Invalid username" and "Invalid password" in all languages using string but the very first toast "Invalid authentication details were passed in api request" still showed then i commented showToast and lLogin and it works now .

**Comment by PratyushSingh07** [comment]
> > Hi @gururani-abhishek that wont be needed all @RobustMj needs to do is change the line 110 of LoginActivity.kt with an appropriate message in all the languages. (Use a Toaster and do not hardcode the string ) However even if you make these changes and allow support for multiple languages you won't be able to check the Toaster message when you switch to any other language as explained in the issue #1997 .
> 
> i added the toast for each of those error "Invalid username" and "Invalid password" in all languages using string but the very first toast "Invalid authentication details were passed in api request" still showed then i commented showToast and lLogin and it works now .

good to know it worked out ,but i suppose your Toaster wont be switching between languages is it?

**Comment by RobustMj** [comment]
> 
@PratyushSingh07 
i think it should switch but we can't check unless #1997 is fixed shouldn't there be another change language option on the login screen?

**Comment by RobustMj** [comment]
@jawidMuhammadi should i raise PR?

**Comment by PratyushSingh07** [comment]
> > 
> 
> @PratyushSingh07 i think it should switch but we can't check unless #1997 is fixed shouldn't there be another change language option on the login screen?

I have checked this in my local system , it will work fine once the PR #2002 is merged

**Comment by RobustMj** [comment]
> > > 
> > 
> > 
> > @PratyushSingh07 i think it should switch but we can't check unless #1997 is fixed shouldn't there be another change language option on the login screen?
> 
> I have checked this in my local system , it will work fine once the PR #2002 is merged

ohk ill just commit then

**Comment by PratyushSingh07** [comment]
> > > > 
> > > 
> > > 
> > > @PratyushSingh07 i think it should switch but we can't check unless #1997 is fixed shouldn't there be another change language option on the login screen?
> > 
> > 
> > I have checked this in my local system , it will work fine once the PR #2002 is merged
> 
> ohk ill just commit then

wait for the mentor to reply

**Comment by RobustMj** [comment]
@PratyushSingh07 ohk

**Comment by RobustMj** [comment]
@PratyushSingh07 why its showing 164 errors in string.xml file?

**Comment by gururani-abhishek** [comment]
> Hi @gururani-abhishek that wont be needed all @RobustMj needs to do is change the line 110 of LoginActivity.kt with an appropriate message in all the languages. (Use a Toaster and do not hardcode the string )
> However even if you make these changes and allow support for multiple languages you won't be able to check the Toaster message when you switch to any other language as explained in the issue #1997 .

That's cool, but I was digging to the root of the issue, right where the exception is received.

**Comment by RobustMj** [comment]
@jawidMuhammadi i ran ./gradlew check in newly downloaded app no changes were made still it showed same errors as before were i made changes .

**Comment by jawidMuhammadi** [comment]
Hey guys, thank you for your valuable input! @RobustMj open a PR and then let's see the whole picture first.

**Comment by preetishsrivastav** [comment]
Hey @jawidMuhammadi is there any test email so that we can login using that

**Comment by PratyushSingh07** [comment]
Hi @preetishsrivastav the credentials are: 
Username : mifos
Password : password

**Comment by ayush290104** [comment]
I am a beginner to open source if anyone can help me to work on this issue..

**Comment by Harsh0707005** [comment]
@PratyushSingh07 @jawidMuhammadi is this resolved?
I wish to work on it

**Comment by PratyushSingh07** [comment]
Yes

**Comment by Harsh0707005** [comment]
@PratyushSingh07 i have request to make
Can you close all the issue that has been resolved.
So it easier for beginners to find an issue to work upon
